### ISSUE TITLE
Null checks not propagating

### ISSUE THREAD

**Issue by ProductiveRage** [Microsoft/CodeContracts#439]
It's been a while since I last looked at code contracts and I wanted to check out where things were up to, so I've started with a very simple example (it might be worth noting that I had an old version installed but I've updated to the latest by downloading the msi from https://github.com/Microsoft/CodeContracts/releases/tag/v1.10.20606.1).

I've started a new solution with a class library "CodeContracts2016" and a console application "Tester". I've enabled code contracts on both projects (ticked "Perform Static Contract Checking" on both but changed no other options aside from setting "Contract Reference Assembly" to "Build" on the class library so that the console application can access the contracts data).

I've created a single class in the class library -

	using System;
	using System.Diagnostics.Contracts;

	namespace CodeContracts2016
	{
		public static class ContractTest
		{
			public static void Hello2(string name)
			{
				Hello(name); // No warning here (I expected one since name isn't checked for non-null)
			}

			public static void Hello(string name)
			{
				Contract.Requires(name != null);
				Console.WriteLine("Hello " + name);
			}
		}
	}

and I call ContractTest.Hello and ContractTest.Hello2 from the console application like this:

	using CodeContracts2016;

	namespace Tester
	{
		class Program
		{
			static void Main(string[] args)
			{
				ContractTest.Hello(null); // Warning here
				ContractTest.Hello2(null); // No warning here
			}
		}
	}

When I build this and wait for the background contract analysis to complete, I expect to get a warning in "Hello2" where I call "Hello" because I've not ensured that "name" is non-null. I get one in "Main", so it's not like it's not working _at all_.

I wondered if the analysis would not flag the call to "Hello" from "Hello2" but, instead, just ensure that "Hello2" was never called with a null value - however, the call to "Hello2" from "Main" does not result in a warning, so it's not that.

Just in case it was possible that there was something causing problems with the cached analysis data (I don't actually know what is cached, so this was clutching at straws), I recreated the solution again from scratch as described above - and encountered exactly the same.

To summarise: I expected a warning when "Hello2" calls "Hello" but didn't get one.

**Comment by yaakov-h** [comment]
Do you have 'Infer Requires' enabled?

**Comment by ProductiveRage** [comment]
I've used all of the defaults - and having just checked; yes, "Infer requires" is ticked. I've attached a screenshot of the options enabled for the class library in case that helps.

![settings](https://cloud.githubusercontent.com/assets/742348/16187332/8e85b424-36c7-11e6-9b35-7c39b47477c4.png)

**Comment by yaakov-h** [comment]
Try turning up your Warning Level.

**Comment by ProductiveRage** [comment]
I turned my warning level all the way up and still didn't get a warning.

If I turn my warnings all the way up and then uncheck "Infer requires" on the class library then I _do_ get the warning.

Is this the correct behaviour? If unchecking "Infer requires" triggers the error in the class library then should _checking_ "Infer requires" be equivalent to

    Contract.Requires(name != null)

being in the "Hello2" method, in which case the "Hello2(null);" call from the console application should have been flagged as a contract not being met?

**Comment by yaakov-h** [comment]
Ah, this mostly makes sense now.

I actually avoid "Infer Requires" because it make things a lot more opaque, as it's doing here. It also seems to be the source of your bug where `ContractTest.Hello2` does not get issued with a potential null reference warning.

Regarding the Console Application, Infer Requires is an option for the static checker, not the rewriter. It will only infer preconditions within the CodeContracts2016 library when it's testing the CodeContracts2016 library.

Those inferences are not done by the rewriter, so CodeContracts2016's Contract Reference Assembly only includes the contract for `ContractTest.Hello`, as it's the only contract that's actually defined.

Thus, you get a potential precondition violation warning when calling `ContractTest.Hello`, but not when calling `ContractTest.Hello2`. `Hello2` does not have any preconditions.

Does that make sense?

**Comment by ProductiveRage** [comment]
That does make sense - I will follow your advice regarding avoiding "Infer Requires", I understand what you're saying about how the Console Application won't infer the same preconditions that the library itself does but it's not something that's immediately obvious.

Are you saying that there is still seemingly a bug around Hello2, though?

**Comment by yaakov-h** [comment]
I don't think there's a bug around `Hello2`. Because it implicitly gets that requires, the precondition on `Hello` is met.

**Comment by ProductiveRage** [comment]
Ok, I understand. Thanks for your help.
